Framework design document (#4)
Mbed-TLS/mbedtls-framework: docs/framework-design.md

Version-independent build and test framework
============================================

## Introduction

The [`mbedtls-framework`](https://github.com/Mbed-TLS/mbedtls-framework) repository provides tooling used to build and test TF-PSA-Crypto (all versions) and Mbed TLS from 3.6.0 onwards.

## Requirements

### Initial motivation

Mbed TLS 3.x was a library for cryptography, X.509 and TLS. In 2024, the cryptography part of the library moved to a separate repository: [TF-PSA-Crypto](https://github.com/Mbed-TLS/TF-PSA-Crypto). Many support files are used by both projects: several helper scripts in `scripts` and `tests/scripts`, most of the test data in `tests/data`, a few helper programs under `programs`, most of the test helper code in `tests/include`, `tests/src` and `tests/drivers`, etc.

> The Mbed TLS project maintains long-time support (LTS) branches (with only bug fixes) in addition to the `development` branch where new features are added. Fixes to bugs often need to be backported from `development` to LTS branches, which involves backporting tests, which often involves backporting test helper code. This would be facilitated by hosting the test helper code in a shared repository that can be consumed by all maintained branches, reducing the amount of code to backport.

Author: Actually I have a problem with your wording. This is in the “Initial motivation” section, and this paragraph explains the old situation. The next paragraph introduces the framework repository. At this point in the explanation, there is no separate repository for helper code: we're explaining why we wanted to create one. I will try to simplify this bit though.

> The Mbed TLS project maintains long-time support (LTS) branches (with only bug fixes) in addition to the `development` branch where new features are added. Fixes to bugs often need to be backported from `development` to LTS branches, which involves backporting tests, which often involves backporting test helper code. This would be facilitated by hosting the test helper code in a shared repository that can be consumed by all maintained branches, reducing the amount of code to backport.

Reviewer: Ah okay, that makes sense. I think my confusion is because it could be read as "this is what we currently do / have done in the past":

"This would [previously] be facilitated by..."

Perhaps a wording more like: "This situation could be improved by hosting..." would be cleaner?

To facilitate parallel development, major changes should avoid breaking existing code and should provide a transition period. For example, if a function needs a new argument, define a new function with a new name, start using the new fuction, and later remove the old function.

### Requirements for consuming repositories

Consuming repositories must have the framework repository as a Git submodule for development work and CI scripts. For merely using the content of the consuming repository, in particular building and running tests, it is enough to have the content of the framework repository, for example from a zip file.

Release archives must include the content of the framework repository.

Without the content of the framework repository, there is no expectation that the consuming repository is usable. In particular, we do not expect the following to work:

* Generating configuration-independent files (e.g. `make generated_files`), not even the ones in the `library` directory.
* `make lib` (with GNU make or CMake) from a pristine checkout.

> * `make test` (even if all the tests have been built).

Reviewer: I think this could be simplified and reworded a bit to make the main points clear:
* Consuming repositories MUST have the framework repo as a submodule
* They MAY rely on the framework git submodule during development
* They MAY rely on the contents of the framework submodule for any build
   * [Subject to previous comment] EXCEPT when configuration-independent generated files are already present, in which case the it MUST be possible to build the library only without the `framework` submodule.
   * So long as the contents match those of the framework commit pointed to by the corresponding submodule in git. Consuming repositories are not required to consume the contents of an arbitrary `framework` version.

Author: I don't really understand your presentation of information, but I think I understand your concern with the current wording, so I've completely reworded this section in a different way.

> We expect `make lib` to work if the generated files in `library` are present and the content of the `framework` directory is missing. This means that in release tags, which include the generated configuration-independent files, we expect `make lib` to work.

Reviewer: Design question: What is the rationale for this requirement? If releases must contain the content of the `framework` repo anyway, we may as well allow consuming repositories to rely on that to build the library.

Unless this is simply a note that it will probably work. If this is the case, it'd be good to note that it is not a requirement for it to work.

Author: Same as today: you can copy `library/*.[hc]` plus `include/**/*.h` and build that, without worrying about the rest of the project.

## CI architecture

Tasks:

* CI in consuming repositories must support Git submodules. Other than that, keep the CI as it is now. In particular, the CI in consuming repositories does not need to consider anything but the commit that the framework submodule points to.
* CI in the framework repository should run a subset of the CI of all consuming repositories, to warn about unintended breakage. This way, most of the time, updating the framework submodule in a consuming repository to the tip of the `main` branch should work. Gatekeepers can bypass this check if the incompatibility is deliberate.
* When merging a pull request to in an official branch in a consuming repository (`development`, LTS branches), check that the commit.

TODO: once this is set up, detail the processes here.

## How to make a change

### Change in a consuming repository requiring a new framework feature

If a change in a consuming repository requires a new feature in the framework, you need to make both a pull request in the framework repository and a pull request in the framework repository.

1. Make a pull request (PR) in the framework repository.

> 2. Upload the framework branch to the framework repository itself (not a fork). This is necessary for the commit to be available on the CI of the consuming repositories (and also for it to be conveniently available to reviewers).

Reviewer: These should be part of the same step or possibly reversed, IIUC. A new branch of the `framework` repository should be created (in the upstream repository, not a fork) and then a pull request should be raised from this branch against `main`.

Author: Not necessarily: for review, the head branch doesn't have to be in the framework repository. That being said, maybe we should make it a policy that in the framework, all pull requests must come from a repository branch? External contributors can't do that though (you need write permission to the repository), which would be a problem sometimes.

This step is only necessary if we want to do CI without more tooling hassle. But we could arrange CI tooling that's capable of using branches from forks (I don't think that's difficult, just not done AFAICT). I think I'm going to leave this section as is, that's an aspect for your subteam to ponder this quarter.

4. If there is rework in the framework PR that is needed for the consuming PR's review or CI, update the framework branch in the framework repository.
5. After the framework PR is merged, update the consuming PR to update the framework submodule to the merge commit (or a later commit).

### Backward-incompatible change in an affected repository

This section discusses cases where a change in the framework repository breaks one or more consuming repositories. This includes cases where the change starts in a consuming repository, for example if some test helper code in the framework repository calls an internal library function which is removed or has its API changed.

#### Split approach for backward-incompatible framework changes

If a change in the framework repository breaks a consuming repository, it should ideally be split into two parts: one that adds the new feature, and one that removes the old feature. The new feature may be gated by a compilation directive if it's convenient to have only one of the versions at compile time.

1. Make and merge a pull request in the framework repository with a backward-compatible change.

> 2. Update all affected consuming repositories to switch to the new version of the framework repository.

Reviewer: I think this clarification is needed:
```suggestion
2. Update all affected consuming repositories to switch to the new version of the framework repository, including switching from the old to the new feature.
```

3. Make and merge a pull request in the framework repository that removes the old version of the feature.

#### Watershed approach for backward-incompatible framework changes

If a change in the framework repository breaks a consuming repository, it is possible to make it in a single step in the framework repository. However, this makes it mandatory to reflect this change in consuming repositories the next time they are updated. Therefore this should only be done if the change can be reflected quickly and there are no other urgent pending framework-submodule updates.

> 1. Make a pull request (PR) in the framework repository with a backward-compatible change.

Reviewer: IIUC this should say "backward-incompatible":
```suggestion
1. Make a pull request (PR) in the framework repository with a backward-incompatible change.
```

> 2. In each affected consuming repository, make a PR that switches to the new version of the framework repository. Wait for those PR to be approved and passing the CI.

Reviewer: Similar clarification:
```suggestion
2. In each affected consuming repository, make a PR that switches to the new version of the framework repository and makes any code changes required for compatibility with the new version. Wait for those PR to be approved and passing the CI.
```

* `test_suite_*.data` contains many test cases that exist in all maintained branches. However the exact expression of these test cases are often different, for example due to compile-time dependencies. Furthermore the set of test cases is often different, for example due to cryptographic mechanisms that are added or removed. Thus we do not expect to share these files. Where there is a lot of commonality, the test cases can be generated from a script located in the framework repository, with code in the generation script to handle the differing parts.

## CI architecture

* CI in consuming repositories must support Git submodules. Other than that, keep the CI as it is now. In particular, the CI in consuming repositories does not need to consider anything but the commit that the framework submodule points to.
* CI in the framework repository should run a subset of the CI of all consuming branches, to warn about unintended breakage. This way, most of the time, updating the framework submodule in a consuming branch to the tip of the `main` branch should work. Gatekeepers can bypass this check if the incompatibility is deliberate.

> * When merging a pull request to in an official branch in a consuming repository (`development`, LTS branches), check that the framework submodule's commit is on the main branch of the `mbedtls-framework` submodule.

Reviewer: Typo:
```suggestion
* When merging a pull request to an official branch in a consuming repository (`development`, LTS branches), check that the framework submodule's commit is on the main branch of the `mbedtls-framework` submodule.
```

However, breaking changes are still problematic. Breaking changes in the framework repository require the affected consuming repositories to fully adapt to the changes when they want to gain access to any new features in the framework repository. Breaking changes in a consuming repository that concern a feature that is consumed by the framework repository (e.g. internal library functions called by test helper functions) have the same effect.

To facilitate parallel development, major changes should avoid breaking existing code and should provide a transition period. For example, if a function needs a new argument, define a new function with a new name, start using the new fuction, and later remove the old function.

### Requirements for consuming repositories

We generalize some current principles:

* For development work and to run the CI, you need a Git checkout.
* To build the project and run functional tests, you need a complete set of files, but you don't need a Git checkout.

> * To just build the library, if the platform-independent generated files are present, you only need the `include` directory and the directories containing library C files (`library`, `3rdparty`, `core`, `drivers` depending on the repository and desired features).

Author: Mostly to simplify the use of the project in other build systems. As noted below in “Requirements for users of consuming repositories”, users who grab a release should just be able to compile `library/*.c` (or equivalent) with just `-I include`.

* `make test` (even if all the tests have been built).

Consuming repositories must not assume that the framework is present when merely building the library. In particular:

* Our provided build scripts (e.g. `library/Makefile`, `library/CMakeLists.txt`) must not require any files from `framework` when compiling the library.
* It's ok to have a file in `library` with a make dependency on a framework file, as long as the build works when the framework file is missing. This allows `make lib` to work as long as the generated files are present.
* Library source files must not rely on headers from the framework.

#### Requirements for users of consuming repositories

Corresponding to the requirements on the repository above:

* Contributors need the framework submodule.
* Users who wish to run full CI tests need the framework submodule.
* Users who want to build or run tests need the `framework` directory content.
* Users who merely want to build the library, and who have the configuration-independent files already generated, do not need the `framework` directory content.

## Contents of the framework repository

### Criteria for inclusion

> In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and should be split into a shared part and a non-shared part.

Author: `tests/suites/test_suite_*` has a lot of code that's shared between branches, but also a lot of code that isn't shared, and tends to have differences inside individual functions and test cases. In particular, 2.28 code has legacy crypto preprocessor guards, 3.6 code has a lot of internal guards like `MBEDTLS_{MD,PK,CIPHER}_{HAVE,CAN}_xxx`, and 4.0 code will have `PSA_WANT`.

We do have places for the parts of `tests/suites/test_suite_*` that can be shared: test generator scripts when a lot of data follows a common model (which usually needs to be fined-tuned to the set of cryptographic mechanisms and API functions available in each branch), and `tests/src/*` for code that is usable in a generic way. This fulfills the “significant proportion of shared content and [MISSING PHRASE] should be split”.

> In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and should be split into a shared part and a non-shared part.

Reviewer 2: Do we gain anything (an increase in how much we can share) if we add a layer of abstraction to allow the guards to be in the separate branches but with the test cases themselves in the framework?

> In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and should be split into a shared part and a non-shared part.

Author: Yes, we do gain some sharing possibilities, at the expense of some complexity. I expect that this complexity will be worth it for some auxiliary test functions. I don't think sharing the unit test cases between branches will be worth it, however. (_Some_ SSL test cases, maybe.) For example, I don't think it would make sense to have RSA test cases in the framework repository, but EdDSA test cases only in TF-PSA-Crypto because older maintained branches don't have EdDSA.

### Requirements for consuming repositories

We generalize some current principles:

* For development work and to run the CI, you need a Git checkout.
* To build the project and run functional tests, you need a complete set of files, but you don't need a Git checkout.
* To just build the library, if the platform-independent generated files are present, you only need the `include` directory and the directories containing library C files (`library`, `3rdparty`, `core`, `drivers` depending on the repository and desired features).

#### Requirements for development in consuming repositories

Consuming repositories must have the framework repository as a Git submodule for development work and CI scripts.

Compared with pre-framework tooling, this means that Git submodules must be enabled. This requires an explicit step in many Git APIs (e.g. running `git submodule update --init` after `git init`, or passing `--recurse-submodules` to `git checkout`).

#### Requirements for processes involving consuming repositories

Release archives must include the content of the framework repository.

#### Requirements for tooling in consuming repositories

> Consuming repositories may assume that the `framework` submodule is present wherever they assume a Git checkout.

Author: Several of `test/scripts/*` run git commands, so they don't work from a release archive anyway. On the other hand, `make && make test && tests/ssl-opt.sh` works from a release archive.

Consuming repositories may assume that the content of the `framework` directory is present anywhere where they would normally assume that all files are present. In particular, this allows the use of framework files for:

* Generating configuration-independent files (e.g. `make generated_files`), including the ones in the `library` directory.
* `make lib` (with GNU make or CMake) from a pristine checkout (because this involves `make generated_files`).

> * `make test` (even if all the tests have been built).

Author: I'm making it explicit that running, and not just building, the tests can require the framework.

Consuming repositories must not assume that the framework is present when merely building the library. In particular:

* Our provided build scripts (e.g. `library/Makefile`, `library/CMakeLists.txt`) must not require any files from `framework` when compiling the library.
* It's ok to have a file in `library` with a make dependency on a framework file, as long as the build works when the framework file is missing. This allows `make lib` to work as long as the generated files are present.
* Library source files must not rely on headers from the framework.

#### Requirements for users of consuming repositories

Corresponding to the requirements on the repository above:

* Contributors need the framework submodule.
* Users who wish to run full CI tests need the framework submodule.
* Users who want to build or run tests need the `framework` directory content.
* Users who merely want to build the library, and who have the configuration-independent files already generated, do not need the `framework` directory content.

## Contents of the framework repository

### Criteria for inclusion

> In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and branch-specific content; such files should be split into a shared part and a non-shared part.

Reviewer 3: ```suggestion
In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming branches. Some files have a significant proportion of shared content and branch-specific content; such files should be split into a shared part and a non-shared part.
``` ?

For example:

* `test_suite_*.function` contains a lot of code that is specific to each branch of each consuming repository. Even when the same test function exists in all maintained branches, there are often minor differences such as a deprecated alternative that only exists in older branches, differences in compile-time dependencies, etc. Thus we do not expect to share these files. Common code can go into separate `.c` files, historically under `tests/src`, that are in the framework repository.
* `test_suite_*.data` contains many test cases that exist in all maintained branches. However the exact expression of these test cases are often different, for example due to compile-time dependencies. Furthermore the set of test cases is often different, for example due to cryptographic mechanisms that are added or removed. Thus we do not expect to share these files. Where there is a lot of commonality, the test cases can be generated from a script located in the framework repository, with code in the generation script to handle the differing parts.

## CI architecture

> Tasks:
> 

Reviewer 3: ```suggestion
```
Just remove this?

* CI in consuming repositories must support Git submodules. Other than that, keep the CI as it is now. In particular, the CI in consuming repositories does not need to consider anything but the commit that the framework submodule points to.

> * CI in the framework repository should run a subset of the CI of all consuming repositories, to warn about unintended breakage. This way, most of the time, updating the framework submodule in a consuming repository to the tip of the `main` branch should work. Gatekeepers can bypass this check if the incompatibility is deliberate.

Reviewer 3: ```suggestion
* CI in the framework repository should run a subset of the CI of all consuming repositories, to warn about unintended breakage. This way, most of the time, updating the framework submodule in a consuming branch to the tip of the `main` branch should work. Gatekeepers can bypass this check if the incompatibility is deliberate.
```

> * When merging a pull request to in an official branch in a consuming repository (`development`, LTS branches), check that the commit.
> 
> TODO: once this is set up, detail the processes here.

Reviewer 3: ```suggestion
```
Just remove this?

Author: Er, why? Both sentences are there for a reason. I will finish the unfinished sentence.

## How to make a change

### Change in a consuming repository requiring a new framework feature

> If a change in a consuming repository requires a new feature in the framework, you need to make both a pull request in the framework repository and a pull request in the framework repository.

Reviewer 3: ```suggestion
If a change in a consuming branch requires a new feature in the framework, you need to make both a pull request in the framework repository and against the consuming branch.
```

1. Make a pull request (PR) in the framework repository.
2. Upload the framework branch to the framework repository itself (not a fork). This is necessary for the commit to be available on the CI of the consuming repositories (and also for it to be conveniently available to reviewers).

>    Open question: can we make the CI work with a fork, and make using forks convenient enough for reviewers, so that people don't need to upload the branch to the main repository?

Reviewer 3: Follow-up on my previous question: `refs/pull/*/head` are commits in the main repository thus by using them the CI would work only with commits in the main repository, right?

Author: They are commits in the main repository, but not fetched by default. On the CI we use Git via a Jenkins plugin and I don't know offhand how flexible it is. I intend to leave this as an open question in this pull request, to be settled by y'all.

> 3. Make a pull request in the consuming repository. Include a commit that advances the submodule to the tip of the branch in the framework repository.

Reviewer 3: ```suggestion
3. Make a pull request against the consuming branch. Include a commit that advances the submodule to the tip of the branch in the framework repository.
```

4. If there is rework in the framework PR that is needed for the consuming PR's review or CI, update the framework branch in the framework repository.
5. After the framework PR is merged, update the consuming PR to update the framework submodule to the merge commit (or a later commit).

> ### Backward-incompatible change in an affected repository

Reviewer 3: ```suggestion
### Backward-incompatible change in the framework repository
```
?

This section discusses cases where a change in the framework repository breaks one or more consuming repositories. This includes cases where the change starts in a consuming repository, for example if some test helper code in the framework repository calls an internal library function which is removed or has its API changed.

#### Split approach for backward-incompatible framework changes

If a change in the framework repository breaks a consuming repository, it should ideally be split into two parts: one that adds the new feature, and one that removes the old feature. The new feature may be gated by a compilation directive if it's convenient to have only one of the versions at compile time.

1. Make and merge a pull request in the framework repository with a backward-compatible change.

> 2. Update all affected consuming repositories to:

Reviewer 3: ```suggestion
2. Update all affected consuming branches to:
```

    1. update the framework submodule to the new version;
    2. migrate all uses of the old feature to the new feature.
3. Make and merge a pull request in the framework repository that removes the old version of the feature.

#### Watershed approach for backward-incompatible framework changes

> If a change in the framework repository breaks a consuming repository, it is possible to make it in a single step in the framework repository. However, this makes it mandatory to reflect this change in consuming repositories the next time they are updated. Therefore this should only be done if the change can be reflected quickly and there are no other urgent pending framework-submodule updates.

Reviewer 3: ```suggestion
If a change in the framework repository breaks a consuming repository, it is possible to make it in a single step in the framework repository. However, this makes it mandatory to reflect this change in consuming branches the next time their framework submodule is updated. Therefore this should only be done if the change can be reflected quickly and there are no other urgent pending framework submodule updates.
```

1. Make a pull request (PR) in the framework repository with a backward-incompatible change.

> 2. In each affected consuming repository, make a PR that updates the framework submodule to the new version and changes the code to work with the updated framework code. Wait for those PR to be approved and passing the CI.

Reviewer 3: ```suggestion
2. For each affected consuming branch, make a PR that updates the framework submodule to the new version and changes the code to work with the updated framework code. Wait for those PR to be approved and passing the CI.
```
